registry/src/api/grpc.rs

```diff
@@ -12,36 +12,110 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::str::FromStr;
+
 use crate::{
     context::Context,
     proto::buffrs::registry::{
         registry_server::Registry, DownloadRequest, DownloadResponse, PublishRequest,
         PublishResponse, VersionsRequest, VersionsResponse,
     },
+    proto::buffrs::package::{
+        Package, Compressed
+    }, types::PackageVersion
 };
 use async_trait::async_trait;
-use tonic::{Request, Response, Status};
+use buffrs::package::PackageName;
+use semver::Version;
+use tonic::{Request, Response, Status, Code};
+use tonic_types::{ErrorDetails, StatusExt};
+
+/// Package's name characters limit, maybe should make it an option later
+const PACKAGE_NAME_LIMIT: usize = 255;
 
 #[async_trait]
 impl Registry for Context {
     async fn publish(
         &self,
-        _request: Request<PublishRequest>,
+        request: Request<PublishRequest>,
     ) -> Result<Response<PublishResponse>, Status> {
-        todo!()
+        let req: PublishRequest = request.into_inner();
+
+        // this shouldn't happen as package isn't optional
+        let package: Compressed = match req.package {
+            None => { return Err(Status::invalid_argument("package wasn't set")); },
+            Some(_value) => {_value}
+        };
+
+        // this shouldn't happen as metadata isn't optional
+        let metadata = match package.metadata {
+            None => { return Err(Status::invalid_argument("metadata wasn't set")); },
+            Some(_value) => {_value} 
+        };
+
+        if metadata.name.len() > PACKAGE_NAME_LIMIT {
+            let mut err_details = ErrorDetails::new();
+            err_details.add_bad_request_violation("name", format!("package's name exceeds limit: {}", PACKAGE_NAME_LIMIT));
+
+            // Generate error status
+            let status = Status::with_error_details(
+                Code::InvalidArgument,
+                "request contains invalid arguments",
+                err_details,
+            );
+            return Err(status)
+        }
+
+        // need to add some more checks (version conflict, do we allow overriding?)
+
+        let storage = self.storage();
+
+        let package_name = PackageName::from_str(metadata.name.as_str());
+        if let Err(error) = package_name {
+            // todo add some logging & proper error handling
+            return Err(Status::invalid_argument(format!("Package name isn't correct, {:?}", error)));
+        }
+
+        let package_version = Version::from_str(metadata.version.as_str());
+        if let Err(_error) = package_name {
+            // todo add some logging & proper error handling
+            return Err(Status::invalid_argument("version isn't correct"));
+        }
+
+        let package_version = &PackageVersion{ package: package_name.unwrap(), version: package_version.unwrap() };
```

Comment:
`.unwrap` is bad practice generally

(if you use the result functions as above, its not needed anymore :))